File: src/Fs.Units/Compound/ElectromotiveForce.fs
namespace Fs.Units

[<AutoOpen>]
module ElectromotiveForce =

  /// Microvolt
  [<Measure>]
  type uV

  /// Millivolt
  [<Measure>]
  type mV

  /// Volt
  [<Measure>]
  type V = J / C

  /// Kilovolt
  [<Measure>]
  type kV

  /// Megavolt
  [<Measure>]
  type MV

  [<RequireQualifiedAccess>]
  module Microvolt =

    let create (x: float) = x * 1.0<uV>
    let toMillivolts (x: float<uV>) = x * 1e-3<mV> / 1.0<uV>
    let toVolts (x: float<uV>) = x * 1e-6<V> / 1.0<uV>
    let toKilovolts (x: float<uV>) = x * 1e-9<kV> / 1.0<uV>
    let toMegavolts (x: float<uV>) = x * 1e-12<MV> / 1.0<uV>

  [<RequireQualifiedAccess>]
  module Millivolt =

    let create (x: float) = x * 1.0<mV>
    let toMicrovolts (x: float<mV>) = x * 1e3<uV> / 1.0<mV>
    let toVolts (x: float<mV>) = x * 1e-3<V> / 1.0<mV>
    let toKilovolts (x: float<mV>) = x * 1e-6<kV> / 1.0<mV>
    let toMegavolts (x: float<mV>) = x * 1e-9<MV> / 1.0<mV>

  [<RequireQualifiedAccess>]
  module Volt =

    let create (x: float) = x * 1.0<V>
    let toMicrovolts (x: float<V>) = x * 1e6<uV> / 1.0<V>
    let toMillivolts (x: float<V>) = x * 1e3<mV> / 1.0<V>
    let toKilovolts (x: float<V>) = x * 1e-3<kV> / 1.0<V>
    let toMegavolts (x: float<V>) = x * 1e-6<MV> / 1.0<V>

  [<RequireQualifiedAccess>]
  module Kilovolt =

    let create (x: float) = x * 1.0<kV>
    let toMicrovolts (x: float<kV>) = x * 1e9<uV> / 1.0<kV>
    let toMillivolts (x: float<kV>) = x * 1e6<mV> / 1.0<kV>
    let toVolts (x: float<kV>) = x * 1e3<V> / 1.0<kV>
    let toMegavolts (x: float<kV>) = x * 1e-3<MV> / 1.0<kV>

  [<RequireQualifiedAccess>]
  module Megavolt =

    let create (x: float) = x * 1.0<MV>
    let toMicrovolts (x: float<MV>) = x * 1e12<uV> / 1.0<MV>
    let toMillivolts (x: float<MV>) = x * 1e9<mV> / 1.0<MV>
    let toVolts (x: float<MV>) = x * 1e6<V> / 1.0<MV>
    let toKilovolts (x: float<MV>) = x * 1e3<kV> / 1.0<MV>

File: src/Fs.Units.CSharp/Compound/ElectromotiveForce.cs
// ReSharper disable once CheckNamespace
namespace Fs.Units.CSharp;

public static class MicrovoltExtensions
{
    public static double MicrovoltToMillivolts(this double x) => Units.ElectromotiveForce.Microvolt.toMillivolts(x);
    public static double MicrovoltToVolts(this double x) => Units.ElectromotiveForce.Microvolt.toVolts(x);
    public static double MicrovoltToKilovolts(this double x) => Units.ElectromotiveForce.Microvolt.toKilovolts(x);
    public static double MicrovoltToMegavolts(this double x) => Units.ElectromotiveForce.Microvolt.toMegavolts(x);
}

public static class MillivoltExtensions
{
    public static double MillivoltToMicrovolts(this double x) => Units.ElectromotiveForce.Millivolt.toMicrovolts(x);
    public static double MillivoltToVolts(this double x) => Units.ElectromotiveForce.Millivolt.toVolts(x);
    public static double MillivoltToKilovolts(this double x) => Units.ElectromotiveForce.Millivolt.toKilovolts(x);
    public static double MillivoltToMegavolts(this double x) => Units.ElectromotiveForce.Millivolt.toMegavolts(x);
}

public static class VoltExtensions
{
    public static double VoltToMicrovolts(this double x) => Units.ElectromotiveForce.Volt.toMicrovolts(x);
    public static double VoltToMillivolts(this double x) => Units.ElectromotiveForce.Volt.toMillivolts(x);
    public static double VoltToKilovolts(this double x) => Units.ElectromotiveForce.Volt.toKilovolts(x);
    public static double VoltToMegavolts(this double x) => Units.ElectromotiveForce.Volt.toMegavolts(x);
}

public static class KilovoltExtensions
{
    public static double KilovoltToMicrovolts(this double x) => Units.ElectromotiveForce.Kilovolt.toMicrovolts(x);
    public static double KilovoltToMillivolts(this double x) => Units.ElectromotiveForce.Kilovolt.toMillivolts(x);
    public static double KilovoltToVolts(this double x) => Units.ElectromotiveForce.Kilovolt.toVolts(x);
    public static double KilovoltToMegavolts(this double x) => Units.ElectromotiveForce.Kilovolt.toMegavolts(x);
}

public static class MegavoltExtensions
{
    public static double MegavoltToMicrovolts(this double x) => Units.ElectromotiveForce.Megavolt.toMicrovolts(x);
    public static double MegavoltToMillivolts(this double x) => Units.ElectromotiveForce.Megavolt.toMillivolts(x);
    public static double MegavoltToVolts(this double x) => Units.ElectromotiveForce.Megavolt.toVolts(x);
    public static double MegavoltToKilovolts(this double x) => Units.ElectromotiveForce.Megavolt.toKilovolts(x);
}